engine/access/ingestion/collection_syncer.go

```
@@ -0,0 +1,398 @@
+	}
+
+	if immediately {
+		s.requester.Force()
+	}
+}
+
+// updateLastFullBlockHeight updates the next highest block height where all previous collections have been indexed.
+// No errors expected during normal operation.
+func (s *CollectionSyncer) updateLastFullBlockHeight() error {
+	lastFullBlockHeight := s.lastFullBlockHeight.Value()
+	lastFinalizedBlock, err := s.state.Final().Head()
+	if err != nil {
+		return fmt.Errorf("failed to get finalized block: %w", err)
+	}
+
+	// track the latest contiguous full height
+	newLastFullBlockHeight, err := s.findLowestBlockHeightWithMissingCollections(lastFullBlockHeight, lastFinalizedBlock.Height)
+	if err != nil {
+		return fmt.Errorf("failed to find last full block height: %w", err)
+	}
+
+	// if more contiguous blocks are now complete, update db
+	if newLastFullBlockHeight > lastFullBlockHeight {
+		err := s.lastFullBlockHeight.Set(newLastFullBlockHeight)
+		if err != nil {
+			return fmt.Errorf("failed to update last full block height: %w", err)
+		}
+
+		s.collectionExecutedMetric.UpdateLastFullBlockHeight(newLastFullBlockHeight)
+
+		s.logger.Debug().
+			Uint64("last_full_block_height", newLastFullBlockHeight).
+			Msg("updated last full block height counter")
+	}
+
+	return nil
+}
+
+// findLowestBlockHeightWithMissingCollections finds the next block height with missing collections,
+// returning the latest contiguous height where all collections are present.
+//
+// No errors expected during normal operations.
+func (s *CollectionSyncer) findLowestBlockHeightWithMissingCollections(
+	lastKnownFullBlockHeight uint64,
+	finalizedBlockHeight uint64,
+) (uint64, error) {
+	newLastFullBlockHeight := lastKnownFullBlockHeight
+
+	for currBlockHeight := lastKnownFullBlockHeight + 1; currBlockHeight <= finalizedBlockHeight; currBlockHeight++ {
+		missingCollections, err := s.findMissingCollectionsAtHeight(currBlockHeight)
+		if err != nil {
+			return 0, ErrMissingCollection
```

similar to `findMissingCollections`, this should just return the error
```suggestion
			return 0, err
```